### PR TITLE
#10309 Define metadata for pagination in REST JSON responses

### PR DESCRIPTION
**Pull request by freddyucv**

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/util/PaginationUtil.java [hunk 1]
```diff
@@ -1,37 +1,215 @@
...
+				ok(new ResponseEntityView(items))
+				.header(LINK_HEADER_NAME, linkHeaderValue)
+				.header(PAGINATION_PER_PAGE_HEADER_NAME, perPageValue)
+				.header(PAGINATION_CURRENT_PAGE_HEADER_NAME, pageValue)
+				.header(PAGINATION_MAX_LINK_PAGES_HEADER_NAME, nLinks)
+				.header(PAGINATION_TOTAL_ENTRIES_HEADER_NAME, totalRecords)
+				.build();
+	}
+
+	private static String getHeaderValue(String urlBase, String filter, int page, int perPage, boolean showArchived,
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/util/PaginationUtil.java]: missing Javadoc.

File: dotCMS/src/main/java/com/dotcms/util/PaginationUtil.java [hunk 2]
```diff
@@ -1,37 +1,215 @@
...
+			links.add(StringUtil.format(LINK_TEMPLATE, map(
+					"URL", getUrl(urlBase, filter, prev, perPage, showArchived, orderBy, direction),
+					"relValue", PREV_REL_VALUE
+			)));
+		}
+
+		return String.join(",", links);
+	}
+
+	private static String getUrl(String urlBase, String filter, int page, int perPage, boolean showArchived,
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/util/PaginationUtil.java]: Missing Javadoc.

File: dotCMS/src/main/java/com/dotcms/util/pagination/Paginator.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import com.liferay.portal.model.User;
+
+import java.util.Collection;
+
+/**
+ * Define the methods to get handle a pagination request
+ */
+public abstract class Paginator<T> {
+
+    public abstract long getTotalRecords(String condition);
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/util/pagination/Paginator.java]: Missing Javadoc.

File: dotCMS/src/main/java/com/dotcms/util/pagination/Paginator.java [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+    public abstract Collection<T> getItems(User user, String filter, boolean showArchived, int limit, int offset,
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/util/pagination/Paginator.java]: Missing Javadoc.

File: dotCMS/src/main/java/com/dotcms/util/pagination/OrderDirection.java [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.dotcms.util.pagination;
+
+public enum OrderDirection {
```
Comment (jcastro-dotcms) [inline dotCMS/src/main/java/com/dotcms/util/pagination/OrderDirection.java]: Missing Javadoc.

File: dotCMS/src/main/java/com/liferay/util/StringUtil.java [hunk 6]
```diff
@@ -652,4 +652,28 @@ public static String wrap(String text, int width, String lineSeparator) {
...
+	 * 	System.out.println(format);
+	 * </code>
+	 *
+	 * The output is: Hello Wordl Freddy
+	 *
+	 * @param template
+	 * @param params
+	 * @return
+	 */
+	public static String format (String template, Map<String, String> params){
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/liferay/util/StringUtil.java]: params must be final

File: dotCMS/src/main/java/com/liferay/util/StringUtil.java [hunk 7]
```diff
@@ -652,4 +652,28 @@ public static String wrap(String text, int width, String lineSeparator) {
...
+		String result = template;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/liferay/util/StringUtil.java]: result should be StringBuilder

File: dotCMS/src/main/java/com/dotcms/rest/api/v1/site/SiteResource.java [hunk 8]
```diff
@@ -115,39 +122,25 @@ public final Response currentSite(@Context final HttpServletRequest req) {
...
+    ) {
 
         Response response = null;
         final InitDataObject initData = this.webResource.init(null, true, req, true, null);
         final User user = initData.getUser();
-        final String filter;
-        final Map<String, Object> paginatedSites;
 
-        try {
+        String filter = (null != filterParam && filterParam.endsWith(NO_FILTER))?
```
Comment (freddyucv) [inline dotCMS/src/main/java/com/dotcms/rest/api/v1/site/SiteResource.java]: Maybe this is something use just for Site

File: dotCMS/src/main/java/com/liferay/util/StringUtil.java (hunk 7 above)
Comment (freddyucv) [inline dotCMS/src/main/java/com/liferay/util/StringUtil.java]: The String:replaceAll method use a StringBuffer

**Review by dsilvam — APPROVED** [review]

**Review by jgambarios — APPROVED** [review]

**Pull request merged**